### DISCUSSION TITLE
How can I get my custom typst template implement the _brand yaml arguments given in the yaml at the top of my qmd document?

### DISCUSSION THREAD

**Discussion by torven-schalk** [quarto-dev/quarto-cli#11616] [Q&A]
### Description

I've used _brand yaml options at the top of my qmd document. but my custom typst template doesn't really want to implement these. Any help on how to solve this would be much appreciated. thanks!
E.g.:

```
---
title: bla
subtitle: bla bla
author: bla bla bla
format:
  my_own_template-typst:
    papersize: a4
brand:
  color:
    primary: "#2C2568"
    secondary: "#107383"
    tertiary: "#37B0A8"
    warning: "#FF4F2D"
  typography:
    fonts:
      - family: Arial
    base:
      family: Arial
      size: "11px"
    link:
      color: secondary
    headings:
      family: Arial
      color: primary
---
```

the headings are correct In Arial and the correct color but the title and normal text are i think the default font.

my typst-template.typ looks like this:

```
#let article(
  title: none,
  subtitle: none,
  authors: none,
  date: none,
  abstract: none,
  abstract-title: none,
  cols: 1,
  margin: (x: 1.25in, y: 1.25in),
  paper: "us-letter",
  lang: "en",
  region: "US",
  font: (),
  fontsize: 11pt,
  sectionnumbering: none,
  toc: false,
  toc_title: none,
  toc_depth: none,
  toc_indent: 1.5em,
  doc,
) = {
  set page(paper: paper, margin: margin)
  set text(font: font, size: fontsize)
  align(horizon + center)[
    #text(size: 24pt, weight: "bold")[#title]\
    
    #text(size: 18pt, weight: "bold")[#subtitle]
    
  ]
  
  align(bottom + left)[#datetime.today().display()]
  pagebreak()
  set page(
    paper: paper,
    margin: margin,
    numbering: "1",
  )
  set par(justify: true)
  set text(lang: lang,
           region: region,
           font: font,
           size: fontsize)
  set heading(numbering: sectionnumbering)

  if title != none {
    align(center)[#block(inset: 2em)[
      #text(weight: "bold", size: 1.5em)[#title]
    ]]
  }

  if authors != none {
    let count = authors.len()
    let ncols = calc.min(count, 3)
    grid(
      columns: (1fr,) * ncols,
      row-gutter: 1.5em,
      ..authors.map(author =>
          align(center)[
            #author.name \
            #author.affiliation \
            #author.email
          ]
      )
    )
  }

  if date != none {
    align(center)[#block(inset: 1em)[
      #date
    ]]
  }

  if abstract != none {
    block(inset: 2em)[
    #text(weight: "semibold")[#abstract-title] #h(1em) #abstract
    ]
  }

  if toc {
    let title = if toc_title == none {
      auto
    } else {
      toc_title
    }
    block(above: 0em, below: 2em)[
    #outline(
      title: toc_title,
      depth: toc_depth,
      indent: toc_indent
    );
    ]
  }

  if cols == 1 {
    doc
  } else {
    columns(cols, doc)
  }
}

#set table(
  inset: 6pt,
  stroke: none
)
```

my typst-show.typ looks like this 

```
#show: doc => article(
$if(title)$
  title: [$title$],
  subtitle: [$subtitle$],
$endif$
$if(by-author)$
  authors: (
$for(by-author)$
$if(it.name.literal)$
    ( name: [$it.name.literal$],
      affiliation: [$for(it.affiliations)$$it.name$$sep$, $endfor$],
      email: [$it.email$] ),
$endif$
$endfor$
    ),
$endif$
$if(date)$
  date: [$date$],
$endif$
$if(lang)$
  lang: "$lang$",
$endif$
$if(region)$
  region: "$region$",
$endif$
$if(abstract)$
  abstract: [$abstract$],
  abstract-title: "$labels.abstract$",
$endif$
$if(margin)$
  margin: ($for(margin/pairs)$$margin.key$: $margin.value$,$endfor$),
$endif$
$if(papersize)$
  paper: "$papersize$",
$endif$
$if(mainfont)$
  font: ("$mainfont$",),
$endif$
$if(fontsize)$
  fontsize: $fontsize$,
$endif$
$if(section-numbering)$
  sectionnumbering: "$section-numbering$",
$endif$
$if(toc)$
  toc: $toc$,
$endif$
$if(toc-title)$
  toc_title: [$toc-title$],
$endif$
$if(toc-indent)$
  toc_indent: $toc-indent$,
$endif$
  toc_depth: $toc-depth$,
  cols: $if(columns)$$columns$$else$1$endif$,
  doc,
)
```

**Comment by mcanouil** [comment]
Did you look at how it's implemented for the default template?

- https://quarto.org/docs/journals/templates.html#typst-partials

**Comment by cderv** [comment]
> the headings are correct In Arial and the correct color but the title and normal text are i think the default font.

This does seems like an issue or current limitation if the font brand setting does not apply yet anywhere 🤔 

The feature is quite new, and it would not the first limitation (e.g. https://github.com/quarto-dev/quarto-cli/issues/11278) 

@gordonwoodhull any insight about font and typst here so that we open the right enhancement if necessary ? 

Thanks !

**Reply by mcanouil** [reply]
I don't think anything from Brand actually works here.
The template here does not use at all the brand and I don't think Quarto is modifying user template.

- https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/typst/pandoc/quarto/typst-show.typ#L38-L42

User created Typst template needs to be updated by users or extension provider to support brand.yml.

**Reply by cderv** [reply]
Oh thank you ! I missed that it was for custom template here... Indeed if an extension provided a custom version of `typst-show.typ` than the brand stuff won't apply. 

This is documentation update for extension writer then. We have a part of using brand in Lua filter (https://quarto.org/docs/authoring/brand.html#lua-api) but not in template for custom format

**Reply by mcanouil** [reply]
Indeed, the other formats rely on SCSS so they are all documented but Typst requires to use the current version of the templates/partials or account for the special brand variable which is not documented (or I missed that part).

The only Typst/brand documentation is how to use brand in content (<https://quarto.org/docs/authoring/brand.html#typst>) but nothing in template/partials.

**Reply by gordonwoodhull** [reply]
I think the essential bug here is that the Typst templates/partials need to be copied into `src/resources/create/format` so that `quarto create extension format:typst` has the latest.

This is the same root cause as #10065 last release.

(I meant to do this, should have opened an issue so as not to forget. Would prefer to automate this so it doesn’t have to be done every release.)

The workaround until I’m able to fix this next week would be to copy the relevant files directly from 

https://github.com/quarto-dev/quarto-cli/tree/main/src/resources/formats/typst/pandoc/quarto

**Reply by mcanouil** [reply]
This would help newly created template indeed but old templates still need to be updated manually by users.
Unfortunately, we don't document this step yet.


For instance the following extension formats needs update to support brand.yml:

- https://github.com/mcanouil/quarto-invoice

or all the ones in https://github.com/quarto-ext/typst-templates

**Reply by gordonwoodhull** [reply]
I don’t think we are able to document how our templates work at that level of detail, sorry.

**Reply by mcanouil** [reply]
I was mostly thinking of documenting the syntax to access brand components

```
$if(brand.typography.headings.family)$
  heading-family: ("$brand.typography.headings.family$",),
$endif$
```

- https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/typst/pandoc/quarto/typst-show.typ#L49-L63

**Reply by gordonwoodhull** [reply]
I see. Yes certainly the way colors are mapped to Typst variables could and should be documented.

Not sure if we should attempt to document the metadata which is passed from Lua filters to the Pandoc templates (like this typography data). That’s a big can of worms. 

In this case it’s using the brand.yml metadata format, but it’s only the base/headings typography part of the brand.yml data that needs to be passed as parameters in`typst-show.typ`, not all of it. It may look like a general facility but currently it is not. 😔

**Reply by cscheid** [reply]
FWIW, I agree with all points that @gordonwoodhull's making here.

**Reply by alinacherkas** [reply]
Hi,

I am facing a similar issue with a custom Typst format. I am following the documentation and `typst-show.typ` which uses `brand.typography` in several places. But it seems that other properties of `brand` are not available. My goal is to grab a logo from `_brand.yml`, if it is defined there, and use it in my template. But here's a much simpler MRE that illustrates the problem:

```yml
# brand.yml
meta:
  name:
    full: Long Name
    short: Short Name
typography:
  fonts:
    - family: Noto Sans
      source: google
  base:
    family: Noto Sans
```

In `typist-show.typ`, I can replace this

```typst
$if(title)$
  title: [$title$],
$endif$
```
with:

```typst
// typst-show.typ
title: [$brand.typography.base.family$],
```

The template will then display  `("Noto Sans",)` as the title. But using this won't display "Long Name":
```typst
title: [$brand.meta.name.full$],
```

@gordonwoodhull , is this what you mean by:

> In this case it’s using the brand.yml metadata format, but it’s only the base/headings typography part of the brand.yml data that needs to be passed as parameters in`typst-show.typ`, not all of it. It may look like a general facility but currently it is not. 😔

I am not sure if I am missing anything or it is only `typography` property that is exposed in `typst` for now. If so, I would be interested to know how to use other properties too.

**Reply by gordonwoodhull** [reply]
Hi @alinacherkas, you are understanding correctly. We do not currently have any general facilities for accessing brand metadata in templates.

Given that templates are the same level of abstraction as brands, it would make sense to expose brand metadata directly to templates. 

But currently you can only access the entire brand metadata from Lua or via the [brand shortcode](https://quarto.org/docs/authoring/brand.html#shortcode---.quarto-shortcode__-escaped-data-is-shortcode1-data-value-brand), neither of which help you with templates. 

The background logo with positioning that we currently support in Typst

https://quarto.org/docs/authoring/brand.html#document-logo-customization

is implemented in this Lua filter:

https://github.com/quarto-dev/quarto-cli/blob/bf038503cd0290ef040288e763174497f5bd7d0b/src/resources/filters/quarto-post/typst-brand-yaml.lua#L217-L305

We have discussed separately that the logo should be implemented as a partial instead of this Lua filter directly generating the 

```typst
#set page(background: ...)
```

Exposing brand metadata to templates would be a similar effort, but more general, because it would expose all of the brand logos to the template, not just the chosen one. 

It could be done either through definition of more constants in Typst [as is done for colors](https://quarto.org/docs/authoring/brand.html#typst-2), or through more metadata exposed to the template as the Typst `brand.typography` metadata discussed here.

**Reply by gordonwoodhull** [reply]
Created #13107 to track this.

**Reply by gordonwoodhull** [reply]
We now have `brand.logo` metadata available as global dictionaries in the Typst output (#13119), and the document logo resolved with brand logo metadata becomes Pandoc metadata interpreted by a new Typst page partial (#13120).

The brand typography data is available through other Pandoc metadata, as shown above.

So I think, except for `brand.meta`, all brand data should be available to Typst templates now. Please comment or open another discussion if this is incorrect.
